tidb-latency-formula.md

```diff
@@ -0,0 +1,531 @@
+# TiDB Latency Formula
+
+This document breaks down the latency into metrics, for those who have interests in observation of the metrics, it's a guide of TiDB's critical path diagnosis.
+
+Generally, OLTP workload can be divided into read and write queries, they share some critical code.
+This document will break down the latency and then analyze it from user's perspective. It's better to read [Performance Analysis and Tuning](/performance-tuning-methods.md) before this document.
+Note when breaking down latency through metrics in this document, we are calculating the average duration or latency instead of some specific slow queries.
+Many metrics are collected as histogram, which is a distribution of the duration or latency, when calculating the average latency, you need to use the sum and count counter.
+
+```
+avg = ${metric_name}_sum / ${metric_name}_count
+```
+
+Metrics in this document can be read directly from prometheus of TiDB.
+
+## General SQL Layer
+
+This part of latency is on the top level of TiDB and shared by any queries.
+
+```
+e2e duration =
+    tidb_server_get_token_duration_seconds +
+    tidb_session_parse_duration_seconds +
+    tidb_session_compile_duration_seconds +
+    tidb_session_execute_duration_seconds{type="general"}
+```
+
+`tidb_server_get_token_duration_seconds` should be small enough to be ignored, usually less than 1 microsecond.
+
+`tidb_session_parse_duration_seconds` records the duration pf parsing sql query text to AST, which can be skipped by `prepare/execute` statements.
+
+`tidb_session_compile_duration_seconds` records the duration of compiling AST to physical plan, which can be skipped by [plan cache](/sql-prepared-plan-cache.md).
+
+`tidb_session_execute_duration_seconds{type="general"}` records the duration of executing duration, which mixes all types of user queries. It need to be broken down into fine-grained durations for analyzing performance issue or bottlenecks.
+
+Now we start dividing read and write queries, they are executed in kind of different ways.
+
+## Read Queries
+
+We start from read queries, which is processed in a single form.
+
+### Point Get
+
+```
+tidb_session_execute_duration_seconds{type="general"} =
+    pd_client_cmd_handle_cmds_duration_seconds{type="wait"} +
+    read handle duration +
+    read value duration
+```
+
+`pd_client_cmd_handle_cmds_duration_seconds{type="wait"}` records the duration of fetching TSO from PD, it will be skipped with in-txn snapshot read or reading with clustered primary index.
```

Comment:
"reading with clustered primary index" and it needs to be an auto commit statement?